File: Programmers/Lv4/Lv4_N-Queen.cpp
#include <string>
#include <vector>

using namespace std;

vector<vector<bool>> arr(13,vector<bool>(13,false));
vector<bool> check_col(13, false);
vector<bool> check_dig(26, false);
vector<bool> check_dig2(26, false);

bool check(int row, int col,int n) {
	// |
	if (check_col[col]) {
		return false;
	}
	// 왼쪽 위 대각선
	if (check_dig[row + col]) {
		return false;
	}
	// /
	if (check_dig2[row - col + n]) {
		return false;
	}
	return true;
}

int calc(int n, int row) {
	if (row == n) {
		// ans += 1;
		return 1;
	}
	int cnt = 0;
	for (int col = 0; col < n; col++) {
		if (check(row, col, n)) {
			check_dig[row + col] = true;
			check_dig2[row - col + n] = true;
			check_col[col] = true;
			arr[row][col] = true;
			cnt += calc(n,row + 1);
			check_dig[row + col] = false;
			check_dig2[row - col + n] = false;
			check_col[col] = false;
			arr[row][col] = false;
		}
	}
	return cnt;
}

int solution111(int n) {
	int answer = calc(n,0);

	return answer;
}

File: Programmers/Lv4/Lv4_N-Queen.py
arr = [[False for x in range(13)] for y in range(13)]
check_col = [False for x in range(13)]
check_dig = [False for x in range(26)]
check_dig2 = [False for x in range(26)]

def check(row, col, n):
    if check_col[col]:      # |
        return False
    if check_dig[row+col]:  # ?쇱そ???媛곸꽑
        return False
    if check_dig2[row-col+n]:   # /
        return False
    return True

def calc(n, row):
    if row == n:
        return 1
    cnt = 0
    for col in range(n):
        if check(row, col, n):
            check_dig[row+col] = True
            check_dig2[row-col+n] = True
            check_col[col] = True
            arr[row][col] = True
            cnt += calc(n, row+1)
            check_dig[row + col] = False
            check_dig2[row - col + n] = False
            check_col[col] = False
            arr[row][col] = False
    return cnt

def solution(n):
    return calc(n,0)